File: backend/src/models/service_health.ts
/**
 * Service Health Interfaces
 * Task: T004 - Create ServiceHealth model/interface
 * Feature: 002-intelligent-message-analysis
 */

export type ServiceStatus = 'online' | 'offline' | 'degraded';

export interface IServiceHealth {
  serviceName: string;
  status: ServiceStatus;
  endpoint?: string;
  responseTime?: number;  // milliseconds
  lastCheckTimestamp: Date;
  errorMessage?: string;
  metadata?: Record<string, any>;
}

export interface IBackendServiceHealth extends IServiceHealth {
  serviceName: 'Backend API';
  endpoint: string;
  responseTime: number;
}

export interface IOllamaServiceHealth extends IServiceHealth {
  serviceName: 'Ollama AI';
  metadata: {
    modelName?: string;
    availableModels?: string[];
    failureCount?: number;
  };
}

export interface IServicesHealthResponse {
  backend: IBackendServiceHealth;
  ollama: IOllamaServiceHealth;
}

File: flutter_app/lib/models/service_health.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'service_health.freezed.dart';
part 'service_health.g.dart';

/// Service Health Model
/// Task: T013 - Create ServiceHealth model
/// Feature: 002-intelligent-message-analysis

enum ServiceStatus {
  online,
  offline,
  degraded,
}

@freezed
class ServiceHealth with _$ServiceHealth {
  const ServiceHealth._();

  const factory ServiceHealth({
    required String serviceName,
    required ServiceStatus status,
    String? endpoint,
    int? responseTime,
    required DateTime lastCheckTimestamp,
    String? errorMessage,
    Map<String, dynamic>? metadata,
  }) = _ServiceHealth;

  factory ServiceHealth.fromJson(Map<String, dynamic> json) =>
      _$ServiceHealthFromJson(json);

  // Computed properties
  bool get isOnline => status == ServiceStatus.online;
  bool get isOffline => status == ServiceStatus.offline;
  bool get isDegraded => status == ServiceStatus.degraded;

  String get statusDisplay {
    switch (status) {
      case ServiceStatus.online:
        return 'En ligne';
      case ServiceStatus.offline:
        return 'Hors ligne';
      case ServiceStatus.degraded:
        return 'Dégradé';
    }
  }

  String get statusIcon {
    switch (status) {
      case ServiceStatus.online:
        return '✅';
      case ServiceStatus.offline:
        return '❌';
      case ServiceStatus.degraded:
        return '⚠️';
    }
  }
}

@freezed
class ServicesHealthResponse with _$ServicesHealthResponse {
  const factory ServicesHealthResponse({
    required bool success,
    required String timestamp,
    required ServicesHealth services,
  }) = _ServicesHealthResponse;

  factory ServicesHealthResponse.fromJson(Map<String, dynamic> json) =>
      _$ServicesHealthResponseFromJson(json);
}

@freezed
class ServicesHealth with _$ServicesHealth {
  const factory ServicesHealth({
    required ServiceHealth backend,
    required ServiceHealth ollama,
  }) = _ServicesHealth;

  factory ServicesHealth.fromJson(Map<String, dynamic> json) =>
      _$ServicesHealthFromJson(json);
}
